dipy/workflows/core.py

```diff
@@ -0,0 +1,426 @@
+
+warnings.filterwarnings('default', module=__name__, message='.*')
+
+BLOCK_TERMINATOR = b'\x00'
+EXTENSION = b'\x21'
+HEADER = b'GIF89a'
+TRAILER = b'\x3b'
+ZERO = b'\x00'
+
+
+def check_dataset_range(dataset):
+    """Confirm no rgb value is outside the range [0, 255]."""
+    if dataset.max() > 255 or dataset.min() < 0:
+        raise ValueError('The dataset has a value outside the range [0,255]')
+
+
+def check_dataset_shape(dataset):
+    """Confirm the dataset has shape 3 x nrows x ncols."""
+    if len(dataset.shape) != 3:
+        raise ValueError('Each image needs 3 dimensions: rgb, nrows, ncols')
+    if dataset.shape[0] != 3:
+        raise ValueError(
+            'The dataset\'s first dimension must have all 3\n'
+            'colors: red, green, and blue...in that order.'
+        )
+
+
+def check_dataset(dataset):
+    """Confirm shape (3 colors x rows x cols) and values [0 to 255] are OK."""
+    if isinstance(dataset, numpy.ndarray) and not len(dataset.shape) == 4:
+        check_dataset_shape(dataset)
+        check_dataset_range(dataset)
+    else:  # must be a list of arrays or a 4D NumPy array
+        for i, d in enumerate(dataset):
+            if not isinstance(d, numpy.ndarray):
+                raise ValueError(
+                    'Requires a NumPy array (rgb x rows x cols) '
+                    'with integer values in the range [0, 255].'
+                )
+            try:
+                check_dataset_shape(d)
+                check_dataset_range(d)
+            except ValueError as err:
+                raise ValueError(
+                    '{}\nAt position {} in the list of arrays.'
+                    .format(err, i)
+                )
+
+
+def try_fix_dataset(dataset):
+    """Transpose the image data if it's in PIL format."""
+    if isinstance(dataset, numpy.ndarray):
+        if len(dataset.shape) == 3:  # NumPy 3D
+            if dataset.shape[-1] == 3:
+                return dataset.transpose((2, 0, 1))
+        elif len(dataset.shape) == 4:  # NumPy 4D
+            if dataset.shape[-1] == 3:
+                return dataset.transpose((0, 3, 1, 2))
+        # Otherwise couldn't fix it.
+        return dataset
+    # List of Numpy 3D arrays.
+    for i, d in enumerate(dataset):
+        if not isinstance(d, numpy.ndarray):
+            return dataset
+        if not (len(d.shape) == 3 and d.shape[-1] == 3):
+            return dataset
+        dataset[i] = d.transpose()
+    return dataset
+
+
+def get_image(dataset):
+    """Convert the NumPy array to two nested lists with r,g,b tuples."""
+    dim, nrow, ncol = dataset.shape
+    uint8_dataset = dataset.astype('uint8')
+    if not (uint8_dataset == dataset).all():
+        message = (
+            "\nYour image was cast to a `uint8` (`<img>.astype(uint8)`), "
+            "but some information was lost.\nPlease check your gif and "
+            "convert to uint8 beforehand if the gif looks wrong."
+        )
+        warnings.warn(message)
```

Comment:
Maybe this is more a question of style, so don't feel forced to change it. I see no advantage in creating a message from a string, then calling a warning with it. Why don't you simply write it in the argument? It takes less place. Idem for this pattern at other places.

    warnings.warn(
        "\nYour image was cast to a `uint8` (`<img>.astype(uint8)`), "
        "but some information was lost.\nPlease check your gif and "
        "convert to uint8 beforehand if the gif looks wrong.")

Also, it's the first time I see a warning starting by `\n`. Is this normal in DiPy?